File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/layout/position/PositionApplier.swift
//
//  PositionApplier.swift
//  StyleEngine/layout/position — Phase 7, step 4 (position).
//
//  Emits per-child SwiftUI modifiers for Position + Inset + ZIndex:
//    • .staticPos              → identity.
//    • .relative with insets   → .offset(x:left, y:top) approximation
//                                 (offset preserves flow, matches CSS).
//    • .absolute               → .offset(x: left ?? -right, y: top ?? -bottom)
//                                 anchored at the positioned ancestor's
//                                 padding-box overlay (the wave-8 mount;
//                                 `needsZStackWrap` flags the container).
//    • .fixed                  → SAME anchored offset, but the box is
//                                 MOUNTED at the canvas root: wave 17's
//                                 FixedHoist strips fixed boxes out of
//                                 their parent tree and hoist-aware
//                                 canvases mount them in FixedHoistOverlay
//                                 at the UNPADDED canvas origin (the
//                                 viewport containing block,
//                                 css-position-3 §3.1). Un-hoisted hosts
//                                 keep the legacy parent anchor verbatim.
//    • .sticky                 → identity + TODO (no SwiftUI equivalent).
//    • zIndex                  → .zIndex(Double(value)).
//
//  Caveats: SwiftUI's .position(x:y:) is coordinate-absolute (centre of
//  the view at the point), which doesn't map cleanly to CSS `top: 0;
//  left: 0` (CSS positions the top-left edge). We use .offset from a
//  ZStack(alignment: .topLeading) so the child's natural frame sits in
//  the upper-left corner and the offset is relative to that corner —
//  which does match CSS exactly for left/top.
//

import SwiftUI
import CoreGraphics

enum PositionApplier {

    /// Apply positioning to a single child view. `agg` is the child's
    /// own LayoutAggregate; `isRTL` comes from the SwiftUI Environment
    /// at the call site.
    static func apply(_ view: AnyView,
                      aggregate agg: LayoutAggregate?,
                      isRTL: Bool = false) -> AnyView {
        // Identity pass when no position fields were set — matches the
        // typography/sizing pattern of "no IR → no modifier overhead".
        guard let agg = agg else { return view }

        // Resolve logical sides against the current layout direction.
        // The extractor defaulted to LTR; flip for RTL here where we can
        // actually see the Environment.
        let rect = agg.inset?.resolved(isRTL: isRTL)

        // Start from the incoming view and layer modifiers. Use AnyView
        // erasures at each step — consistent with the other engine
        // appliers in the project that chain into a single return path.
        var out = view

        // Wave 49 (lane A7) — the PERCENTAGE-INSET lane. A percentage inset
        // resolves against the containing block's corresponding dimension
        // (CSS 2.1 §9.4.3), a number only the SwiftUI Environment carries
        // (`containingBlockWidth` / `containingBlockHeight`), which this
        // static func cannot read. A ViewModifier can — the same trick
        // BorderSideApplier uses for its own `@Environment(\.wptCaptureMode)`
        // read — so the geometry switch below moves into [insetGeometry] and
        // the percentage case routes through it from inside composition.
        // Entered ONLY when a side really was declared as a percentage, so
        // every px-inset component keeps the pre-wave-49 view tree exactly.
        if let pct = agg.percentInsets, pct.any {
            out = AnyView(view.modifier(PercentInsetPositioned(
                kind: agg.position ?? .staticPos,
                legacyRect: rect,
                percents: pct.resolved(isRTL: isRTL))))
            if let z = agg.zIndex { out = AnyView(out.zIndex(z)) }
            return out
        }

        out = insetGeometry(view, kind: agg.position ?? .staticPos, rect: rect)

        // zIndex applies regardless of positioning scheme — CSS allows
        // z-index on any positioned element, and SwiftUI's .zIndex works
        // on any view (paint-order tie-breaker at the parent stack).
        if let z = agg.zIndex {
            out = AnyView(out.zIndex(z))
        }

        return out
    }

    /// The position-scheme geometry, split out of [apply] (wave 49) so the
    /// percentage lane can run the IDENTICAL switch on its resolved rect —
    /// the two can never drift apart.
    fileprivate static func insetGeometry(_ view: AnyView,
                                          kind: PositionKind,
                                          rect: InsetRect?) -> AnyView {
        var out = view
        switch kind {
        case .staticPos:
            // No offset. Identity (zIndex handled below).
            break

        case .relative:
            // Relative offset keeps the element in flow and nudges its
            // paint by (left, top). Right/Bottom are ignored when the
            // corresponding opposite side is present — matches CSS.
            if let r = rect {
                let x = r.left ?? (r.right.map { -$0 } ?? 0)
                let y = r.top  ?? (r.bottom.map { -$0 } ?? 0)
                if x != 0 || y != 0 {
                    out = AnyView(out.offset(x: x, y: y))
                }
            }

        case .absolute:
            // F2 (css-position-3 §3.1): an absolute box anchors at its
            // containing block's corner + insets. The renderer mounts it
            // in the positioned ancestor's padding-box overlay (the
            // wave-8 anchor, wave-9/11 containing-block channels), or —
            // for a ROOT-level box — a hoist-aware canvas mounts it in
            // the canvas-root FixedHoistOverlay (wave 17: the initial
            // containing block is the UNPADDED canvas, per the measured
            // web behavior). Either way THIS modifier only expresses
            // "corner of my anchoring ZStack + inset offset".
            out = anchoredInsetOffset(out, rect: rect)

        case .fixed:
            // F1 (wave 17, css-position-3 §3.1): a fixed box's containing
            // block is the VIEWPORT — never the parent. The containing-
            // block difference from .absolute is established by WHERE the
            // box is MOUNTED, not by a different offset formula:
            // FixedHoist strips fixed boxes out of their parent's tree
            // and a hoist-aware canvas (the composed WPT canvas) mounts
            // them in the canvas-root FixedHoistOverlay ABOVE its
            // padding, so this identical anchored offset resolves
            // against the unpadded canvas origin — insets as absolute
            // anchors, no parent flow slot, no compounding (pin S3).
            // A host that renders WITHOUT the hoist (the dark-stage
            // per-component canvas, plain SDUI trees) keeps the legacy
            // parent-anchored geometry byte-identically — the
            // 327-baseline gate — because the math below is shared.
            out = anchoredInsetOffset(out, rect: rect)

        case .sticky:
            // SwiftUI has no sticky modifier; the closest thing is a
            // ScrollView with a pinned header. Log a TODO and render
            // identity so the rest of the test suite still passes.
            // TODO: integrate with the sticky-header TOC PR once it lands.
            break
        }

        return out
    }

    /// The shared out-of-flow geometry (wave 17 split of the old fused
    /// `.absolute, .fixed` case — same math, now cited per scheme at the
    /// call sites): anchor the box at the corner of its mounting ZStack
    /// picked by WHICH insets are declared, then offset by them.
    ///
    /// left / top → straight offset from the top-leading corner.
    /// right / bottom without left / top → anchor at the trailing/bottom
    /// edge and offset NEGATIVELY. Since we have no container size at
    /// this layer, the flexible frame + alignment stands in for true
    /// edge resolution. TODO: replace with GeometryReader-based
    /// right/bottom inset math when a corpus test pins it.
    ///
    /// A nil rect (no insets at all) is IDENTITY: the box stays at its
    /// static position (CSS 2.1 §10.3.7/§10.6.4 — all-auto insets),
    /// which the renderer's static-position shifts then own.
    private static func anchoredInsetOffset(_ view: AnyView,
                                            rect r: InsetRect?) -> AnyView {
        // No insets → static position; nothing for this modifier to do.
        guard let r = r else { return view }
        // Any declared side on an axis claims that axis's anchor.
        let hasLeftOrRight = (r.left != nil) || (r.right != nil)
        let hasTopOrBottom = (r.top != nil)  || (r.bottom != nil)
        // All four auto (empty rect object) → identity, same as nil.
        guard hasLeftOrRight || hasTopOrBottom else { return view }
        // Pick the anchoring corner per axis: a declared left/top wins
        // (CSS resolves over-constrained insets in favor of the
        // inline/block start in LTR — CSS 2.1 §10.3.7 rule order).
        let h: HorizontalAlignment = (r.left != nil) ? .leading
                                    : (r.right != nil) ? .trailing
                                    : .leading
        let v: VerticalAlignment = (r.top != nil) ? .top
                                 : (r.bottom != nil) ? .bottom
                                 : .top
        // Offset magnitude: start insets push positively; end insets
        // (right/bottom) pull NEGATIVELY from the trailing anchor.
        let dx = r.left ?? (r.right.map { -$0 } ?? 0)
        let dy = r.top  ?? (r.bottom.map { -$0 } ?? 0)
        return AnyView(
            view
                // Flow-filling frame so we can anchor to a corner of
                // whatever ZStack mounts this box (parent overlay for
                // absolute, canvas-root FixedHoistOverlay for hoisted
                // fixed/absolute roots).
                .frame(maxWidth: .infinity,
                       maxHeight: .infinity,
                       alignment: Alignment(horizontal: h, vertical: v))
                .offset(x: dx, y: dy)
        )
    }

    /// Returns true when at least one child of a container carries
    /// position: absolute | fixed — signalling the container must be a
    /// ZStack(alignment: .topLeading) so offsets resolve from the
    /// top-left corner of the parent's bounds.
    ///
    /// Callers: ComponentRenderer inspects children's parsed aggregates
    /// before choosing a container shape.
    static func needsZStackWrap(forChildren childAggregates: [LayoutAggregate?]) -> Bool {
        for agg in childAggregates {
            guard let p = agg?.position else { continue }
            if p == .absolute || p == .fixed { return true }
        }
        return false
    }

    /// Wave 49 (lane A7) — the composed half of the percentage-inset lane.
    ///
    /// A `ViewModifier` is the only place in this engine that can read the
    /// containing-block environment channels, because `PositionApplier.apply`
    /// is a static function called from `ComponentRenderer`'s body and gets
    /// no Environment of its own. Same pattern as `BorderSideApplier`'s
    /// `@Environment(\.wptCaptureMode)` read.
    ///
    /// WPT-capture gated for the reason the spacing appliers gate: the
    /// committed baselines contain percent insets too
    /// (fixtures/properties/layout/position-top-left.json,
    /// inset-logical.json, fidelity/layout.combos.json) and were frozen
    /// against the number-as-points reading. Outside WPT capture the legacy
    /// rect is used verbatim, so those baselines are byte-identical by
    /// construction.
    fileprivate struct PercentInsetPositioned: ViewModifier {
        /// The parent-published containing-block content width; nil when the
        /// ancestor's inline size is not statically definite.
        @Environment(\.containingBlockWidth) private var cbWidth
        /// Block-axis twin. Nil means indefinite (Renderer/ContainingBlock).
        @Environment(\.containingBlockHeight) private var cbHeight
        /// Only the WPT capture lane takes the resolved value (see above).
        @Environment(\.wptCaptureMode) private var wptCaptureMode

        let kind: PositionKind
        let legacyRect: InsetRect?
        let percents: InsetPercents

        func body(content: Content) -> some View {
            let rect = wptCaptureMode
                ? PercentInsetResolve.resolve(legacy: legacyRect,
                                              percents: percents,
                                              cbWidth: cbWidth,
                                              cbHeight: cbHeight)
                : legacyRect
            return PositionApplier.insetGeometry(AnyView(content), kind: kind, rect: rect)
        }
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/layout/position/PositionApplier.kt
package com.styleconverter.runtime.layout.position

import androidx.compose.foundation.layout.absoluteOffset
import androidx.compose.ui.Modifier
// The value form of the offset this applier emits — see [resolvedOffset],
// which the backdrop draw node reads because it sits OUTER of this step.
import androidx.compose.ui.unit.DpOffset
import androidx.compose.ui.unit.dp
import androidx.compose.ui.zIndex

/**
 * Applies position configuration to a Compose Modifier.
 *
 * ## CSS Position Mapping
 *
 * | CSS Position | Compose Equivalent |
 * |--------------|-------------------|
 * | static       | No modifier (default) |
 * | relative     | `Modifier.absoluteOffset()` (CSS insets are physical) |
 * | absolute     | `Box` + `Modifier.absoluteOffset()` (container handling required) |
 * | fixed        | `Modifier.absoluteOffset()` from a [CanvasRootHoist] overlay anchor |
 * | sticky       | [StickyPositionWrapper] with scroll-aware behavior |
 *
 * ## Implementation Notes
 *
 * 1. **Absolute positioning**: In CSS, `position: absolute` removes the element from
 *    the normal flow and positions it relative to the nearest positioned ancestor.
 *    In Compose, this requires the parent to be a `Box` and uses
 *    `Modifier.absoluteOffset()` (physical CSS insets, css-position-3 §3.1).
 *    The [needsAbsoluteContainer] method indicates when this is needed. With NO
 *    positioned ancestor the containing block is the initial containing block —
 *    the capture canvas — and [CanvasRootHoist] re-anchors the box there
 *    (wave 17, pin S2); the wave-8/9 positioned-container machinery keeps
 *    owning the positioned-ancestor case (pin S4).
 *
 * 2. **Fixed positioning** (wave 17): CSS `position: fixed` anchors at the
 *    VIEWPORT (css-position-3 §3.2) — our capture canvas. [CanvasRootHoist]
 *    hoists the box out of its parent's flow entirely and renders it from a
 *    zero-size overlay anchor at the UNPADDED canvas origin; the FIXED branch
 *    below then applies the (left, top) inset as an absoluteOffset from that
 *    origin. The former Popup-based FixedPositionWrapper was deleted as dead
 *    code — do NOT resurrect the Popup route: a Popup renders into a SEPARATE
 *    WINDOW that never composites into the capture bitmap (PixelCopy and the
 *    composed canvas's GraphicsLayer record only the canvas's own window
 *    surface), so a Popup-fixed box is invisible to every capture pipeline.
 *
 * 3. **Sticky positioning**: CSS `position: sticky` is a hybrid that acts like
 *    relative until a scroll threshold, then acts like fixed. Implemented via
 *    [StickyPositionWrapper] which tracks scroll state and applies dynamic offset.
 *
 * ## z-index
 *
 * CSS `z-index` maps directly to `Modifier.zIndex()` for controlling
 * draw order within the same parent.
 */
object PositionApplier {

    /**
     * Apply position configuration to a modifier.
     *
     * @param modifier The base modifier to extend
     * @param config The position configuration to apply
     * @return Modified modifier with positioning applied
     */
    fun applyPosition(modifier: Modifier, config: PositionConfig): Modifier {
        if (!config.hasPosition) return modifier

        // Wave 49 (lane A7) — the PERCENTAGE-INSET lane. A percentage inset
        // resolves against the containing block's corresponding dimension
        // (CSS 2.1 §9.4.3), a number only composition can see: it rides
        // `LocalContainingBlock`, which this static chain cannot read. The
        // composed detour therefore lives in its own file (the per-file size
        // rule) — [percentInsetPositioned] — and is entered ONLY when a side
        // was actually declared as a percentage, so every px-inset component
        // keeps the pre-wave-49 modifier chain exactly.
        val pct = config.percent
        if (pct != null && pct.any) return modifier.percentInsetPositioned(config)

        return applyResolvedPosition(modifier, config)
    }

    /**
     * The z-index + offset chain for a config whose insets are already
     * absolute Dp values. Split out of [applyPosition] so the percentage
     * lane above can run the identical chain on its resolved copy — the
     * two can never drift apart.
     */
    internal fun applyResolvedPosition(modifier: Modifier, config: PositionConfig): Modifier {
        var result = modifier

        // Apply z-index for stacking order
        if (config.hasZIndex) {
            result = result.zIndex(config.zIndex)
        }

        // Apply positioning based on type
        result = when (config.type) {
            PositionType.STATIC -> {
                // Static elements have no offset
                result
            }

            PositionType.RELATIVE -> {
                // Relative: offset from normal position
                applyOffset(result, config)
            }

            PositionType.ABSOLUTE -> {
                // Absolute: positioned relative to container
                // Note: This requires the parent to be a Box with proper alignment
                applyOffset(result, config)
            }

            PositionType.FIXED -> {
                // Fixed (wave 17): the containing block is the VIEWPORT —
                // our capture canvas (css-position-3 §3.2). The offset below
                // is the (left, top) INSET from wherever this node is
                // anchored: under CanvasRootHoist.Host (the composed WPT
                // canvas) the node renders from a zero-size overlay anchor
                // at the UNPADDED canvas origin, so this lands the box at
                // canvas (left, top) exactly like the Chromium reference
                // (pins S1/S3 — the parent's flow slot contributes NOTHING;
                // the pre-wave-17 parent-relative reading of this same
                // modifier was the diagnosed bug). On hostless paths (the
                // dark-stage per-component canvas, whose wave-1 out-of-flow
                // root contract already anchors a root-level subject) the
                // chain is byte-identical to the frozen 327-pair baseline.
                applyOffset(result, config)
            }

            PositionType.STICKY -> {
                // Sticky: hybrid behavior
                // Limited support - treat as relative for basic offset
                applyOffset(result, config)
            }
        }

        return result
    }

    /**
     * The offset [applyPosition] will actually emit for [config], as a value
     * instead of a modifier — the exact analogue of
     * `MarginApplier.resolvedInsets`.
     *
     * ## Why this exists (wave-27 backdrop fix)
     * `StyleApplier` installs the effects step (step 3) OUTSIDE the layout
     * step (step 4), and this applier's `absoluteOffset` is the last thing
     * step 4 chains. A draw node at step 3 is therefore OUTER of the offset:
     * its `positionInWindow()` and its local draw origin both describe the
     * element's UN-offset slot, while the element's own background/borders
     * (steps 5–6, INNER of the offset) paint at the offset one.
     *
     * The backdrop lane is the one consumer that has to reconcile the two —
     * it samples the canvas at a window coordinate AND paints into that draw
     * space, so a dropped offset mis-places both halves *consistently* (the
     * patch looked like a perfect filter of the wrong rectangle: measured on
     * `backdrop-filter-basic.html`, the inverted patch landed at the parent
     * colorbox's origin, exactly the child's `left:50px; top:50px` short).
     *
     * Returning the value from the same `config.offsetX`/`offsetY` reads
     * [applyOffset] uses — behind the same `hasPosition`/STATIC gates — means
     * the backdrop node and the layout chain can never disagree about how far
     * the box moved. Zero for a static (or offsetless) element, which is the
     * overwhelming majority and costs them nothing.
     *
     * KNOWN LIMIT (wave 49, lane A7) — stated rather than hidden, and the
     * exact shape MarginApplier.resolvedInsets already documents for its own
     * percent lane: a PERCENTAGE inset ([PositionConfig.percent]) is resolved
     * only inside [percentInsetPositioned]'s composed lane, which this plain
     * function cannot reproduce (it has no CompositionLocal scope). A percentage-inset
     * element that ALSO carries backdrop-filter / filter / box-shadow /
     * mix-blend-mode would therefore have its effect painted at the legacy
     * number-as-pixels offset while its own paint sits at the resolved one.
     * No corpus test combines the two: the only bare-number (percentage)
     * insets in all 30 frozen wave-48 sections are css-position/
     * position-relative-001…006 and -008, none of which declare an effect.
     * The fix when one arrives is to thread the resolved offset out of the
     * composed lane, exactly as the margin note prescribes.
     */
    fun resolvedOffset(config: PositionConfig): DpOffset {
        // Gate 1 — mirrors applyPosition's own first line: no positioning
        // properties at all means no modifier was chained, so no offset.
        if (!config.hasPosition) return DpOffset.Zero
        // Gate 2 — mirrors the STATIC branch of applyPosition's `when`: a
        // static box ignores its insets entirely (css-position-3 §2), and is
        // the only type that does NOT route through applyOffset.
        if (config.type == PositionType.STATIC) return DpOffset.Zero
        // Every other type (relative / absolute / fixed / sticky) calls
        // applyOffset, which reads exactly these two accessors.
        return DpOffset(config.offsetX, config.offsetY)
    }

    /**
     * Apply offset based on resolved top/right/bottom/left values.
     *
     * This applies the calculated offsetX and offsetY from the config,
     * which already handles precedence rules (left over right, top over bottom).
     */
    private fun applyOffset(modifier: Modifier, config: PositionConfig): Modifier {
        // Wave 22 (B-RC3): these offsets are SIGNED insets from whichever
        // anchor the box's declared sides select — positive from the start
        // (left/top) anchor, NEGATIVE from the end (right/bottom) anchor
        // (css-position-3 §3.5.3). Choosing the anchor is deliberately NOT
        // this modifier's job: it has no containing-block extent and no
        // measured box size. The mounting slot owns it — CanvasRootHoist's
        // overlay anchors an end-only-inset box flush with the canvas edge
        // (cb − box) and this modifier's `−right` then pulls it inward to
        // the CSS used position cb − box − right. iOS composes the same two
        // halves (PositionApplier.anchoredInsetOffset: a `.trailing`/
        // `.bottom` frame alignment plus the identical negated offset), so
        // the two natives share one rule table. Where no slot resolves an
        // extent, the start anchor is kept — EndInsetAnchor's A4.
        //
        // Wave 28 (lane NE): there are TWO such slots now. The canvas-hoist
        // overlay owns ROOT-level boxes; a box with a POSITIONED ANCESTOR is
        // mounted by [PositionedAncestorAnchor] instead (the wave-8/9
        // RenderAbsoluteChild path, which anchored every child at the slot
        // origin and so painted `right:20; bottom:20` at (−20,−20)). Both
        // slots delegate to the same EndInsetAnchor arithmetic, so this
        // modifier's half of the composition is unchanged for either.
        val x = config.offsetX
        val y = config.offsetY

        // Only apply offset if there's actual movement
        return if (x.value != 0f || y.value != 0f) {
            // Wave 12: absoluteOffset, NOT offset. CSS top/right/bottom/left
            // are PHYSICAL insets (css-position-3 §3.1: `left` always
            // measures from the physical LEFT edge — only the logical
            // inset-inline-* spellings are direction-aware, and PositionConfig
            // already resolved those into the physical slots). This applier is
            // the LIVE render chain (StyleApplier → LayoutFacade →
            // applyPosition) and used the layout-direction-aware
            // Modifier.offset(), so under Direction:RTL (ComponentRenderer
            // provides LocalLayoutDirection=Rtl) a CSS left:-20px mirrored to
            // +20px physical: the three rtl abspos-autopos containers
            // pixel-measured at x56 instead of x16 (+40px total error).
            modifier.absoluteOffset(x = x, y = y)
        } else {
            modifier
        }
    }

    /**
     * Apply only z-index without any offset.
     * Useful when offset is handled separately (e.g., in container logic).
     *
     * @param modifier The base modifier
     * @param config The position configuration
     * @return Modifier with only z-index applied
     */
    fun applyZIndexOnly(modifier: Modifier, config: PositionConfig): Modifier {
        return if (config.hasZIndex) {
            modifier.zIndex(config.zIndex)
        } else {
            modifier
        }
    }

    /**
     * Check if this component needs special container handling for absolute/fixed positioning.
     *
     * When true, the parent container should:
     * 1. Use `Box` as the container type
     * 2. Position this child using alignment and offset
     * 3. Handle the element being "out of flow"
     *
     * @param config The position configuration to check
     * @return True if absolute container handling is needed
     */
    fun needsAbsoluteContainer(config: PositionConfig): Boolean {
        return config.isAbsolutelyPositioned
    }

    /**
     * Check if a property type is handled by this applier.
     */
    fun isPositionProperty(propertyType: String): Boolean {
        return PositionExtractor.isPositionProperty(propertyType)
    }

    // needsFixedWrapper / needsPositionWrapper were deleted in wave 17
    // together with the Popup-based FixedPositionWrapper they routed to
    // (dead code — no call sites; see the fixed-positioning note in the
    // header for why the Popup route must never come back). Fixed boxes now
    // ride CanvasRootHoist + the FIXED branch above.

    /**
     * Check if this config requires a StickyPositionWrapper.
     *
     * When true, the component should be wrapped with [StickyPositionWrapper]
     * instead of using normal modifier-based positioning.
     */
    fun needsStickyWrapper(config: PositionConfig): Boolean {
        return config.type == PositionType.STICKY
    }
}
